### PR TITLE
Refactor/core dictlist2

### PR DESCRIPTION
**Pull request by akaviaLab**
Refactor dictlist.py and test_dictlist.py to be Python 3.6+ compliant including type hints, f-strings, no utf-8, no future imports

### CODE REVIEW

File: src/cobra/core/dictlist.py [hunk 1]
```diff
@@ -26,7 +27,7 @@ def __init__(self, *args):
 
         """
         if len(args) > 2:
-            raise TypeError("takes at most 1 argument (%d given)" % len(args))
+            raise TypeError(f"takes at most 1 argument ({len(args):d} given)")
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py:41]: The `:d` format specification is not strictly necessary, I think, but it also doesn't hurt.

File: src/cobra/core/dictlist.py [hunk 2]
```diff
@@ -37,32 +38,33 @@ def __init__(self, *args):
             else:
                 self.extend(other)
 
-    def has_id(self, id):
+    def has_id(self, id: Any) -> bool:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]: Pretty sure `id` has to be a string, no?
```suggestion
    def has_id(self, id: str) -> bool:
```

Comment (Midnighter) [reply src/cobra/core/dictlist.py]: Or is it `Union[cobra.core.Object, str]`?

File: src/cobra/core/dictlist.py [hunk 3]
```diff
@@ -37,32 +38,33 @@ def __init__(self, *args):
...
+        """Check if id is in DictList."""
         return id in self._dict
 
-    def _check(self, id):
-        """make sure duplicate id's are not added.
+    def _check(self, id: Any) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]: Same question about `id`.

File: src/cobra/core/dictlist.py [hunk 4]
```diff
@@ -37,32 +38,33 @@ def __init__(self, *args):
...
 
-    def _generate_index(self):
-        """rebuild the _dict index"""
+    def _generate_index(self) -> None:
+        """Rebuild the _dict index."""
         self._dict = {v.id: k for k, v in enumerate(self)}
 
-    def get_by_id(self, id):
-        """return the element with a matching id"""
+    def get_by_id(self, id: Any) -> Any:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]: Same, what is `id`'s type?

File: src/cobra/core/dictlist.py [hunk 5]
```diff
@@ -122,7 +128,7 @@ def query(self, search_function, attribute=None):
         >>> model.metabolites.query(regex, attribute='name')
         """
 
-        def select_attribute(x):
+        def select_attribute(x: Union[type(None), Any]) -> Any:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
        def select_attribute(x: Optional[Any]) -> Any:
```

File: src/cobra/core/dictlist.py [hunk 6]
```diff
@@ -150,35 +154,41 @@ def select_attribute(x):
...
+        """Append object to end."""
+        the_id = entity.id
         self._check(the_id)
         self._dict[the_id] = len(self)
-        list.append(self, object)
+        list.append(self, entity)
 
-    def union(self, iterable):
-        """adds elements with id's not already in the model"""
+    def union(self, iterable: Iterable) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]: Is this too specific?

```suggestion
    def union(self, iterable: Iterable[Object]) -> None:
```

File: src/cobra/core/dictlist.py [hunk 7]
```diff
@@ -150,35 +154,41 @@ def select_attribute(x):
...
+        will not have initialized yet, because the initialization class was
+        left unspecified. This is an issue because unpickling calls
+        DictList.extend, which requires the presence of _dict. Therefore,
+        the issue is caught and addressed here.
+
+        Parameters
+        ----------
+        iterable : Iterable
+        """
         if not hasattr(self, "_dict") or self._dict is None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
        if getattr(self, "_dict", None) is None:
```

File: src/cobra/core/dictlist.py [hunk 8]
```diff
@@ -150,35 +154,41 @@ def select_attribute(x):
...
                 append(i)
 
-    def extend(self, iterable):
-        """extend list by appending elements from the iterable"""
-        # Sometimes during initialization from an older pickle, _dict
-        # will not have initialized yet, because the initialization class was
-        # left unspecified. This is an issue because unpickling calls
-        # DictList.extend, which requires the presence of _dict. Therefore,
-        # the issue is caught and addressed here.
+    def extend(self, iterable: Iterable) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]: I think

```suggestion
    def extend(self, iterable: Iterable[Object]) -> None:
```

File: src/cobra/core/dictlist.py [hunk 9]
```diff
@@ -195,72 +205,86 @@ def extend(self, iterable):
...
                 raise ValueError(
-                    "id '%s' at index %d is non-unique. "
-                    "Is it present twice?" % (str(the_id), i)
+                    f"id '{str(the_id)}' at index {i :d} is non-unique. "
+                    f"Is it present twice?"
                 )
 
-    def _extend_nocheck(self, iterable):
-        """extends without checking for uniqueness
+    def _extend_nocheck(self, iterable: Iterable) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def _extend_nocheck(self, iterable: Iterable[Object]) -> None:
```

File: src/cobra/core/dictlist.py [hunk 10]
```diff
@@ -195,72 +205,86 @@ def extend(self, iterable):
...
-        if current_length is 0:
+        if not current_length:
             self._generate_index()
             return
         for i, obj in enumerate(islice(self, current_length, None), current_length):
             _dict[obj.id] = i
 
-    def __sub__(self, other):
-        """x.__sub__(y) <==> x - y
+    def __sub__(self, other: Iterable) -> "DictList":
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def __sub__(self, other: Iterable[Object]) -> "DictList":
```

File: src/cobra/core/dictlist.py [hunk 11]
```diff
@@ -195,72 +205,86 @@ def extend(self, iterable):
...
         """
         total = DictList()
         total.extend(self)
         for item in other:
             total.remove(item)
         return total
 
-    def __isub__(self, other):
-        """x.__sub__(y) <==> x -= y
+    def __isub__(self, other: Iterable) -> "DictList":
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def __isub__(self, other: Iterable[Object]) -> "DictList":
```

File: src/cobra/core/dictlist.py [hunk 12]
```diff
@@ -195,72 +205,86 @@ def extend(self, iterable):
...
             other must contain only unique id's present in the list
         """
-
         for item in other:
             self.remove(item)
         return self
 
-    def __add__(self, other):
-        """x.__add__(y) <==> x + y
+    def __add__(self, other: Iterable) -> "DictList":
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def __add__(self, other: Iterable[Object]) -> "DictList":
```

File: src/cobra/core/dictlist.py [hunk 13]
```diff
@@ -195,72 +205,86 @@ def extend(self, iterable):
...
-
         """
         total = DictList()
         total.extend(self)
         total.extend(other)
         return total
 
-    def __iadd__(self, other):
-        """x.__iadd__(y) <==> x += y
+    def __iadd__(self, other: Iterable) -> "DictList":
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def __iadd__(self, other: Iterable[Object]) -> "DictList":
```

File: src/cobra/core/dictlist.py [hunk 14]
```diff
@@ -272,28 +296,37 @@ def __iadd__(self, other):
...
         This is only provided for backwards compatibility so older
         versions of cobrapy can load pickles generated with cobrapy. In
-        reality, the "_dict" state is ignored when loading a pickle"""
+        reality, the "_dict" state is ignored when loading a pickle
+        """
         return {"_dict": self._dict}
 
-    def __setstate__(self, state):
-        """sets internal state
+    def __setstate__(self, state) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def __setstate__(self, state: dict) -> None:
```

File: src/cobra/core/dictlist.py [hunk 15]
```diff
@@ -302,52 +335,57 @@ def index(self, id, *args):
...
         the_copy = DictList()
         list.extend(the_copy, self)
         the_copy._dict = self._dict.copy()
         return the_copy
 
-    def insert(self, index, object):
-        """insert object before index"""
-        self._check(object.id)
-        list.insert(self, index, object)
+    def insert(self, index, entity) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def insert(self, index: int, entity: Object) -> None:
```

File: src/cobra/core/dictlist.py [hunk 16]
```diff
@@ -302,52 +335,57 @@ def index(self, id, *args):
...
         _dict = self._dict
         for i, j in _dict.items():
             if j >= index:
                 _dict[i] = j + 1
-        _dict[object.id] = index
+        _dict[entity.id] = index
 
-    def pop(self, *args):
-        """remove and return item at index (default last)."""
+    def pop(self, *args) -> Any:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def pop(self, *args) -> Object:
```

File: src/cobra/core/dictlist.py [hunk 17]
```diff
@@ -357,24 +395,28 @@ def pop(self, *args):
                 _dict[i] = j - 1
         return value
 
-    def add(self, x):
-        """Opposite of `remove`. Mirrors set.add"""
+    def add(self, x: Any) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def add(self, x: Object) -> None:
```

File: src/cobra/core/dictlist.py [hunk 18]
```diff
@@ -357,24 +395,28 @@ def pop(self, *args):
...
+        """Opposite of `remove`. Mirrors set.add."""
         self.extend([x])
 
-    def remove(self, x):
-        """.. warning :: Internal use only"""
-        # Each item is unique in the list which allows this
-        # It is much faster to do a dict lookup than n string comparisons
+    def remove(self, x: Any) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def remove(self, x: Object) -> None:
```

File: src/cobra/core/dictlist.py [hunk 19]
```diff
@@ -388,7 +430,10 @@ def key(i):
 
         self._generate_index()
 
-    def __getitem__(self, i):
+    def __getitem__(
+        self, i: Union[int, slice, Iterable, Any]
+    ) -> Union["DictList", Any]:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
        self, i: Union[int, slice, Iterable, Object]
    ) -> Union["DictList", Object]:
```

File: src/cobra/core/dictlist.py [hunk 20]
```diff
@@ -406,7 +451,16 @@ def __getitem__(self, i):
         else:
             return list.__getitem__(self, i)
 
-    def __setitem__(self, i, y):
+    def __setitem__(self, i: Union[slice, int], y: Union[list, Any]) -> None:
```
Comment (Midnighter) [inline src/cobra/core/dictlist.py]:
```suggestion
    def __setitem__(self, i: Union[slice, int], y: Union[list, Object]) -> None:
```

**Review by Midnighter — CHANGES_REQUESTED** [review]
Thank you for updating the module. I'm unfortunately not too willing to allow too many `Any` annotations since that doesn't add helpful information. I think, in most cases the `Any` can be replaced with either `Object` or `Union[Object, str]`.

**Comment by akaviaLab** [comment]
> Thank you for updating the module. I'm unfortunately not too willing to allow too many `Any` annotations since that doesn't add helpful information. I think, in most cases the `Any` can be replaced with either `Object` or `Union[Object, str]`.

I was a bit confused by get_by_any, which allows multiple types. I wasn't sure what dictlist allows. I can turn everything to Union[str, Object] and then we'll see.

**Comment by Midnighter** [comment]
When the parameter is `id` it can often be `Object` or `str`.  I can take another careful look when you're done and just commit the changes to your branch.

**Comment by akaviaLab** [comment]
Yes, please do so. I trust you, and this code is confusing to me.

**Review by Midnighter — APPROVED** [review]
Good stuff. We can go over some of the annotations at a later point, should there be issues with them.

**Pull request merged**